ngraph_bridge/ngraph_executor.cc

```diff
@@ -0,0 +1,550 @@
+      static_input_map[i] = &input_tensor;
+      TF_RETURN_IF_ERROR(TensorToStream(signature_ss, input_tensor));
+      signature_ss << ";";
+    }
+  }
+  return Status::OK();
+}
+
+//---------------------------------------------------------------------------
+//  NGraphExecutor::GetNgExecutable
+//---------------------------------------------------------------------------
+Status NGraphExecutor::GetNgExecutable(
+    const std::vector<Tensor>& tf_input_tensors,
+    std::shared_ptr<ngraph::runtime::Executable>& ng_exec, bool& cache_hit) {
+  // FIRST Compute the signature
+  std::stringstream signature_ss;
+  string signature;
+
+  std::shared_ptr<ngraph::Function> ng_function;
+  std::shared_ptr<ngraph::runtime::Executable> evicted_ng_exec;
+
+  NGRAPH_VLOG(4) << "GetNgExecutable: Got backend of type: "
+                 << m_op_backend_name;
+
+  // Get the backend. Note that the backend may not be available
+  // so that's a programmng error.
+  ng::runtime::Backend* op_backend;
+  try {
+    op_backend = BackendManager::GetBackend(m_op_backend_name);
+  } catch (...) {
+    return errors::Internal("Backend not available: ", m_op_backend_name);
+  }
+
+  lock_guard<mutex> lock(m_mutext);
+
+  std::vector<TensorShape> input_shapes;
+  std::vector<const Tensor*> static_input_map;
+  TF_RETURN_IF_ERROR(ComputeSignature(tf_input_tensors, input_shapes,
+                                      static_input_map, signature_ss));
+  signature = signature_ss.str();
+
+  NGRAPH_VLOG(5) << "Computed signature: " << signature;
+
+  cache_hit = false;
+  auto it = m_ng_exec_map.find(signature);
+
+  // Translate the TensorFlow graph to nGraph.
+  if (it == m_ng_exec_map.end()) {
+    // Measure the current total memory usage
+    long vm, rss, vm0, rss0;
+    MemoryProfile(vm0, rss0);
+
+    NGRAPH_VLOG(1) << "Compilation cache miss: " << m_name;
+    if (!m_do_aot) {
+      TF_RETURN_IF_ERROR(Builder::TranslateGraph(input_shapes, static_input_map,
+                                                 m_graph.get(), ng_function));
+      ng_function->set_friendly_name(m_name);
+    } else {
+      auto itr = m_aot_functions.find(signature);
+      if (itr == m_aot_functions.end()) {
+        return errors::Internal(
+            "Expected to find AOT precompiled ng function of signature: ",
+            signature);
+      }
+      ng_function = ng::deserialize(itr->second);
+    }
+
+    auto function_size = ng_function->get_graph_size() / 1024;  // kb unit
+
+    // Serialize to nGraph if needed
+    if (std::getenv("NGRAPH_ENABLE_SERIALIZE") != nullptr) {
+      std::string file_name = "tf_function_" + m_name + ".json";
+      NgraphSerialize("tf_function_" + m_name + ".json", ng_function);
+#if defined NGRAPH_DISTRIBUTED
+      int rank_id;
+      rank_id = ng::get_distributed_interface()->get_rank();
+      NgraphSerialize(
+          "tf_function_" + m_name + "_" + to_string(rank_id) + ".json",
+          ng_function);
+#endif
```

Comment:
free this function of input output caches.  Is this in scope of this PR?

Reply:
Yes - but a different PR